src/web/tokenserver.rs

```diff
@@ -1,26 +1,162 @@
+        &Validation::new(Algorithm::RS256),
+    ).map_err(|ee| {
+        ApiError::from(ApiErrorKind::Internal(format!("Unable to decode token_data: {:}", ee)))
+    })?;
+    let email = format!("{:}@api.accounts.firefox.com", token_data.claims.sub);
+
+    // TODO pull out of settings instead
+    let shared_secret = env::var("SYNC_MASTER_SECRET").expect("SYNC_MASTER_SECRET must be set");
+    let database_url =
+        env::var("TOKENSERVER_DATABASE_URL").expect("TOKENSERVER_DATABASE_URL must be set");
+
+    let connection = MysqlConnection::establish(&database_url)
+        .unwrap_or_else(|_| panic!("Error connecting to {}", database_url));
+    let user_record = diesel::sql_query(
+        "select users.uid, services.pattern, users.email, users.generation, users.client_state, users.created_at, users.replaced_at, nodes.node, users.keys_changed_at from users, services, nodes where users.email = ? and services.id = users.service and nodes.id = users.nodeid and nodes.service = services.id")
+        .bind::<Text, _>(email)
+        .load::<TokenserverUser>(&connection).unwrap();
+    let (python_result, python_derived_result) = Python::with_gil(|py| {
+        let tokenlib = PyModule::from_code(
+            py,
+            r#"
+import tokenlib
+
+def make_token(plaintext, shared_secret):
+    return tokenlib.make_token(plaintext, secret=shared_secret)
+
+def get_derived_secret(plaintext, shared_secret):
+    return tokenlib.get_derived_secret(plaintext, secret=shared_secret)
+"#,
+            "main.py",
+            "main",
+        )
+        .map_err(|e| {
+            e.print_and_set_sys_last_vars(py);
+            e
+        })?;
+        let thedict = [
+            ("node", user_record[0].node.as_ref()),
+            ("uid", token_data.claims.sub.as_ref()),
+            ("fxa_kid", "asdf"), // userid component of authorization email
+            ("fxa_uid", "qwer"),
+            ("hashed_device_id", "..."),
+            ("hashed_fxa_uid", "..."),
+        ]
+        .into_py_dict(py);
+        // todo don't hardcode
+        // we're supposed to check the "duration" query
+        // param and use that if present (for testing)
+        thedict.set_item("expires", 300).unwrap();
+        let result = match tokenlib.call1("make_token", (thedict, &shared_secret)) {
+            Err(e) => {
+                e.print_and_set_sys_last_vars(py);
+                return Err(e);
+            }
+            Ok(x) => x.extract::<String>().unwrap(),
+        };
+        let derived_result = match tokenlib.call1("get_derived_secret", (&result, &shared_secret)) {
+            Err(e) => {
+                e.print_and_set_sys_last_vars(py);
+                return Err(e);
+            }
+            Ok(x) => x.extract::<String>().unwrap(),
+        };
+        //assert_eq!(result, false);
+        Ok((result, derived_result))
+    })
+    .unwrap();
+    let api_endpoint = format!("{:}/1.5/{:}/", user_record[0].node, token_data.claims.sub);
```

Comment:
We want the uid from the db here instead of the fxa id. +nit: i don't think this usually renders a trailing slash

```suggestion
    let api_endpoint = format!("{:}/1.5/{:}", user_record[0].node, user_record[0].uid);
```